### ISSUE TITLE
Memory friendly changes. (avoid crashes on lower memory machines)

### ISSUE THREAD

**Issue by dtorrero** [rtorrero/TuneTranslator-model#3]
# Memory Issue Fix Summary

## Problem Identified

Your training script crashed with an out-of-memory (OOM) error because it attempts to **preload the entire dataset into RAM** at initialization. With ~2,740 training songs and your current configuration, this exceeds your 64GB RAM.

## Root Causes

1. **Full Dataset Preloading**: The `MelTabDataset` class loads all mel-spectrograms and tokens into memory
2. **Large Batch Size**: `BATCH_SIZE = 128` (now reduced to 32)
3. **Unlimited Segments**: `MAX_SEGMENTS_PER_SONG = None` means all measures from each song are loaded
4. **High Resolution**: `N_MELS = 256` and `MAX_SEGMENT_FRAMES = 200` create large tensors
5. **Multiple DataLoader Workers**: 4 workers with `persistent_workers=True` duplicate memory

### 1. Reduced Batch Size
- **Before**: `BATCH_SIZE = 128`
- **After**: `BATCH_SIZE = 32`
- **Impact**: 4x reduction in batch memory usage

### 2. Limited Segments Per Song
- **Before**: `MAX_SEGMENTS_PER_SONG = None` (unlimited)
- **After**: `MAX_SEGMENTS_PER_SONG = 20`
- **Impact**: Caps dataset size, especially for long songs

### 3. Reduced DataLoader Workers
- **Before**: `num_workers=4`
- **After**: `num_workers=2`
- **Impact**: Reduces worker process memory overhead

### 4. Added Memory Monitoring
- Added system memory check at training start
- Warns if less than 20GB available
- Requires `psutil` package (install with: `pip install psutil`)

## Estimated Memory Usage

With the new settings:
- **Training segments**: ~2,740 songs × 20 segments = ~54,800 segments
- **Validation segments**: ~343 songs × 20 segments = ~6,860 segments
- **Per segment**: ~256 mels × 200 frames × 4 bytes = ~200KB
- **Total dataset**: ~54,800 × 200KB ≈ **11GB**
- **Model + optimizer**: ~5-10GB
- **Batch buffers**: ~2-4GB
- **Total estimated**: **~20-25GB** (should fit in 64GB with headroom)

### If Still Running Out of Memory

1. **Further reduce batch size**:
   ```python
   BATCH_SIZE = 16  # or even 8
   ```

2. **Reduce segments per song**:
   ```python
   MAX_SEGMENTS_PER_SONG = 10  # or even 5
   ```

3. **Reduce model size**:
   ```python
   D_MODEL = 512      # from 768
   LAYERS = 6         # from 8
   ```

4. **Use lazy loading** (requires code changes):
   - Modify `MelTabDataset` to load data on-demand instead of preloading
   - Trade speed for memory efficiency

### Monitor Memory During Training

Install and use `psutil`:
```bash
pip install psutil
```

Check memory usage:
```python
import psutil
mem = psutil.virtual_memory()
print(f"Used: {mem.percent}%, Available: {mem.available / (1024**3):.1f} GB")
```

### Optimize Gradient Accumulation

Since batch size is reduced, increase gradient accumulation to maintain effective batch size:
```python
GRADIENT_ACCUMULATION_STEPS = 4  # Effective batch size = 32 × 4 = 128
```

This is already set to 2 in your config, consider increasing to 4 or 8.

## Testing the Fix

Run training again:
```bash
python train.py
```

The script will now:
1. Show memory status before loading
2. Load a limited dataset (~20-25GB instead of 60GB+)
3. Use smaller batches to prevent OOM during training

## Long-term Solution

Consider implementing a **lazy-loading dataset** that loads data on-demand:
- Loads mel-spectrograms from disk as needed
- Uses an LRU cache to keep frequently accessed data in memory
- Trades some speed for much better memory efficiency
- Allows training on datasets larger than available RAM

This would require modifying the `MelTabDataset` class but would eliminate OOM issues entirely.

## Questions?

If you still experience OOM errors:
1. Check actual memory usage with `htop` or `nvidia-smi` (for GPU)
2. Further reduce `BATCH_SIZE` or `MAX_SEGMENTS_PER_SONG`
3. Consider using a smaller subset of your data for initial experiments
